src/UnitCell.cpp

```diff
@@ -9,261 +9,254 @@
+    }
+
+    bool is_roughly_90(double value) {
         // We think that 89.999° is close enough to 90°
-        return fabs(angle - 90.0) < 1e-3;
-    };
-    if (is_zero(a_) && is_zero(b_) && is_zero(c_)) {
-        shape_ = INFINITE;
-        a_ = b_ = c_ = 0;
-        alpha_ = beta_ = gamma_ = 90;
-    } else if (is_90(alpha_) && is_90(beta_) && is_90(gamma_)) {
-        shape_ = ORTHORHOMBIC;
-        // Make sure alpha/beta/gamma are actually 90°, so that the matrix
-        // update below does not create a non diagonal matrix.
-        alpha_ = 90;
-        beta_ = 90;
-        gamma_ = 90;
-    } else {
-        shape_ = TRICLINIC;
+        return fabs(value - 90.0) < 1e-3;
     }
-    update_matrix();
-}
 
-UnitCell::UnitCell(const Matrix3D& matrix): h_(Matrix3D::unit()), h_inv_(Matrix3D::unit()) {
-    if (matrix[1][0] != 0 || matrix[2][0] != 0 || matrix[2][1] != 0) {
-        throw error("the matrix supplied to UnitCell is not an upper triangular matrix");
+    bool is_diagonal(const Matrix3D& matrix) {
+        return is_roughly_zero(matrix[1][0]) && is_roughly_zero(matrix[2][0]) &&
+               is_roughly_zero(matrix[0][1]) && is_roughly_zero(matrix[2][1]) &&
+               is_roughly_zero(matrix[0][2]) && is_roughly_zero(matrix[1][2]);
     }
 
-    if (matrix[0][0] == 0 && matrix[1][1] == 0 && matrix[2][2] == 0 &&
-        matrix[0][1] == 0 && matrix[0][2] == 0 && matrix[1][2] == 0) {
+    bool is_upper_triangular(const Matrix3D& matrix) {
+        return is_roughly_zero(matrix[1][0]) && is_roughly_zero(matrix[2][0]) &&
+               is_roughly_zero(matrix[2][1]);
+    }
+}}
 
-        shape_ = INFINITE;
-        a_ = b_ = c_ = 0;
-        alpha_ = beta_ = gamma_ = 90.0;
+using namespace chemfiles::private_details;
 
-        update_matrix();
+static void check_lengths(const Vector3D& lengths) {
+    if (lengths[0] < 0 || lengths[1] < 0 || lengths[2] < 0) {
+        throw error("a unit cell can not have negative lengths");
+    }
 
-        return;
+    if (lengths != Vector3D(0, 0, 0) && (is_roughly_zero(lengths[0]) || is_roughly_zero(lengths[1]) || is_roughly_zero(lengths[2]))) {
+        warning("", "trying to set one or two unit cell lengths to zero, something might be wrong");
     }
+}
 
-    if (matrix[0][1] == 0 && matrix[0][2] == 0 && matrix[1][2] == 0) {
-        shape_ = ORTHORHOMBIC;
+static void check_angles(const Vector3D& angles) {
+    if (angles[0] < 0 || angles[1] < 0 || angles[2] < 0) {
+        throw error("a unit cell can not have negative angles");
+    }
 
-        a_ = matrix[0][0];
-        b_ = matrix[1][1];
-        c_ = matrix[2][2];
+    if (is_roughly_zero(angles[0]) || is_roughly_zero(angles[1]) || is_roughly_zero(angles[2])) {
+        throw error("a unit cell can not have 0° angles");
+    }
 
-        alpha_ = beta_ = gamma_ = 90.0;
+    if (angles[0] > 180 || angles[1] > 180 || angles[2] > 180) {
+        throw error("a unit cell can not have angles larger than 180°");
+    }
```

Comment:
This would allow nonphysical angles of exactly 180°. Would `>=` be better? I'm pretty indifferent about that, so fell free to decide however you want.

Reply:
Sure, that might help a bit =)

Reply:
One very last nitpick: it would probably be nice to have the error say "... larger than or equal to 180°".
But then this is good to go :D